### PR TITLE
File api

### PR DESCRIPTION
**Pull request by jekanik**

### CODE REVIEW

File: application/src/main/java/com/musicforall/config/WebAppConfig.java [hunk 1]
```diff
@@ -13,20 +13,28 @@
...
+        VelocityConfigurer conf = new VelocityConfigurer();
+        conf.setResourceLoaderPath("/WEB-INF/velocity/");
+        return conf;
+    }
+
+    @Bean
+    public CommonsMultipartResolver multipartResolver() {
+        CommonsMultipartResolver resolver = new CommonsMultipartResolver();
+        resolver.setDefaultEncoding("utf-8");
+        resolver.setMaxUploadSizePerFile(50000000);
```
Comment (cat-eat-dog) [inline application/src/main/java/com/musicforall/config/WebAppConfig.java]: make max upload size to be configurable via properties

File: application/src/main/java/com/musicforall/web/FileController.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * @author Evgeniy on 11.06.2016.
+ */
+
+@Controller
+public class FileController {
```
Comment (cat-eat-dog) [inline application/src/main/java/com/musicforall/web/FileController.java]: - [ ]  make more like rest api files/ POST to upload, files/{fileName} GET to retrieve file
- [ ]  all file-specific logic should be moved to file manager (directory creation, exception handling, etc)

File: file-api/src/main/java/com/musicforall/files/FileManager.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+public class FileManager {
+
+    private String workingDirectory;
+
+    public FileManager(final String directory) {
+        workingDirectory = directory;
+    }
+
+    public boolean saveMultipart(final MultipartFile file) throws IOException {
```
Comment (cat-eat-dog) [inline file-api/src/main/java/com/musicforall/files/FileManager.java]: rename to save

File: file-api/src/main/java/com/musicforall/files/FileManager.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * @author Evgeniy on 11.06.2016.
+ */
+
+public class FileManager {
+
+    private String workingDirectory;
+
+    public FileManager(final String directory) {
```
Comment (cat-eat-dog) [inline file-api/src/main/java/com/musicforall/files/FileManager.java]: directory must be autowired and configured via properties files

File: file-api/src/main/java/com/musicforall/files/FileManager.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        workingDirectory = directory;
+    }
+
+    public boolean saveMultipart(final MultipartFile file) throws IOException {
+        long savedBytes = Files.copy(file.getInputStream(), Paths.get(workingDirectory, file.getOriginalFilename()));
+        return savedBytes == file.getSize();
+    }
+
+    public void streamFileByName(final String name, final OutputStream destination) throws IOException {
```
Comment (cat-eat-dog) [inline file-api/src/main/java/com/musicforall/files/FileManager.java]: method must return File

File: file-api/src/main/java/com/musicforall/files/FileManager.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        Stream<Path> walk = Files.walk(Paths.get(workingDirectory), 2);
+        List<Path> paths = walk
+                .filter(path -> path.toString().toLowerCase().contains(name.toLowerCase()))
+                .collect(Collectors.toList());
+        if (paths.isEmpty()) return;
```
Comment (cat-eat-dog) [inline file-api/src/main/java/com/musicforall/files/FileManager.java]: use FileSystems.getDefault().getPath(workingDirectory, name) instead

File: file-api/src/test/java/com/musicforall/files/FileManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    private static FileManager manager;
+    private static final URL resourceUrl = FileManagerTest.class.getClassLoader().getResource("test_resource.jpg");
+
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        String projectDitPath = System.getProperty("user.dir");
+        String testDirPath = projectDitPath + File.separator + "src" + File.separator + "test" + File.separator + "for_test" + File.separator;
+        testDirectory = new File(testDirPath);
+        testDirectory.mkdirs();
```
Comment (cat-eat-dog) [inline file-api/src/test/java/com/musicforall/files/FileManagerTest.java]: move to utils class

File: file-api/src/main/java/com/musicforall/files/FileManager.java (hunk 6 above)
Comment (jekanik) [inline file-api/src/main/java/com/musicforall/files/FileManager.java]: done

File: file-api/src/main/java/com/musicforall/files/FileManager.java (hunk 3 above)
Comment (jekanik) [inline file-api/src/main/java/com/musicforall/files/FileManager.java]: done

File: file-api/src/main/java/com/musicforall/files/FileManager.java (hunk 5 above)
Comment (jekanik) [inline file-api/src/main/java/com/musicforall/files/FileManager.java]: done

File: application/src/main/java/com/musicforall/config/WebAppConfig.java (hunk 1 above)
Comment (jekanik) [inline application/src/main/java/com/musicforall/config/WebAppConfig.java]: done

File: application/src/main/java/com/musicforall/web/FileController.java (hunk 2 above)
Comment (jekanik) [inline application/src/main/java/com/musicforall/web/FileController.java]: done

File: application/pom.xml [hunk 8]
```diff
@@ -24,7 +23,11 @@
             <version>1.1.0</version>
         </dependency>
 
-        <!-- todo add file api dependency-->
+        <dependency>
+            <groupId>com.music.for.all</groupId>
+            <artifactId>file-api</artifactId>
+            <version>1.1.0</version>
```
Comment (cat-eat-dog) [inline application/pom.xml]: use maven property to specify version

File: file-api/src/main/java/com/musicforall/files/manager/FileManager.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            savedBytes = Files.copy(file.getInputStream(), Paths.get(workingDirectory, file.getOriginalFilename()));
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return savedBytes == file.getSize();
+    }
+
+    public Path getFileByName(final String fileName) {
+        Path path = Paths.get(workingDirectory, fileName);
+        if (Files.exists(path)) return null;
```
Comment (cat-eat-dog) [inline file-api/src/main/java/com/musicforall/files/manager/FileManager.java]: check if statement

Comment (jekanik) [inline file-api/src/main/java/com/musicforall/files/manager/FileManager.java]: done

File: file-api/src/main/java/com/musicforall/files/manager/FileManager.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        File dir = new File(workingDirectory);
+        dir.mkdirs();
+    }
+
+    public boolean save(final MultipartFile file) {
+        long savedBytes = 0;
+        try {
+            savedBytes = Files.copy(file.getInputStream(), Paths.get(workingDirectory, file.getOriginalFilename()));
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (cat-eat-dog) [inline file-api/src/main/java/com/musicforall/files/manager/FileManager.java]: return false if exception happened

Comment (jekanik) [inline file-api/src/main/java/com/musicforall/files/manager/FileManager.java]: done

File: application/pom.xml (hunk 8 above)
Comment (jekanik) [inline application/pom.xml]: done

File: file-api/src/main/java/com/musicforall/files/FileManager.java (hunk 4 above)
Comment (jekanik) [inline file-api/src/main/java/com/musicforall/files/FileManager.java]: done

File: application/src/main/java/com/musicforall/web/FileController.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+    @RequestMapping(value = "/files/{fileName}", method = RequestMethod.GET)
+    public void getFileHandler(HttpServletRequest request, HttpServletResponse response, @PathVariable("fileName") String name) {
+        Path filePath = manager.getFileByName(name);
+        if (filePath == null) return;
+
+        try {
+            Files.copy(filePath, response.getOutputStream());
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (cat-eat-dog) [inline application/src/main/java/com/musicforall/web/FileController.java]: use logger instead

File: file-api/src/main/java/com/musicforall/files/FileApiSpringConfig.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+@Configuration
+@ComponentScan("com.musicforall.files.manager")
+public class FileApiSpringConfig {
+
+    @Bean
+    public CommonsMultipartResolver multipartResolver() {
+        CommonsMultipartResolver resolver = new CommonsMultipartResolver();
+        resolver.setDefaultEncoding("utf-8");
+//        resolver.setMaxUploadSizePerFile(Long.valueOf(env.getRequiredProperty("web.max_upload_size_per_file")));
+        resolver.setMaxUploadSizePerFile(50000000);
```
Comment (cat-eat-dog) [inline file-api/src/main/java/com/musicforall/files/FileApiSpringConfig.java]: should be configured via properties

File: file-api/src/test/java/com/musicforall/files/FileManagerTest.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+import static org.junit.Assert.assertTrue;
+
+public class FileManagerTest {
+
+    private final URL resourceUrl = FileManagerTest.class.getClassLoader().getResource("test_resource.jpg");
+    private File testDirectory;
+    private FileManager manager;
+
+    @Before
+    public void setUp() throws Exception {
```
Comment (cat-eat-dog) [inline file-api/src/test/java/com/musicforall/files/FileManagerTest.java]: test directory must be created in BeforeClass

File: file-api/src/main/java/com/musicforall/files/FileApiSpringConfig.java (hunk 12 above)
Comment (jekanik) [inline file-api/src/main/java/com/musicforall/files/FileApiSpringConfig.java]: done

File: file-api/src/test/java/com/musicforall/files/FileManagerTest.java (hunk 13 above)
Comment (jekanik) [inline file-api/src/test/java/com/musicforall/files/FileManagerTest.java]: done

File: application/src/main/java/com/musicforall/web/FileController.java (hunk 11 above)
Comment (jekanik) [inline application/src/main/java/com/musicforall/web/FileController.java]: done

File: file-api/src/test/java/com/musicforall/files/FileManagerTest.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.file.Path;
+
+import static java.nio.file.Files.*;
+import static java.nio.file.Paths.get;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class FileManagerTest {
```
Comment (cat-eat-dog) [inline file-api/src/test/java/com/musicforall/files/FileManagerTest.java]: also need to check io exception

Comment (jekanik) [inline file-api/src/test/java/com/musicforall/files/FileManagerTest.java]: done

**Pull request merged**